BIP324 connection support

sipa opened the pull request:
This is part of #27634.

This implements the BIP324 v2 transport (which implements all of what the BIP calls transport layer *and* application layer), though in a non-exposed way. It is tested through an extensive fuzz test, which verifies that v2 transports can talk to v2 transports, and v1 transports can talk to v2 transports, and a unit test that exercises a number of unusual scenarios. The transport is functionally complete, including:
* Autodetection of incoming V1 connections.
* Garbage, both sending and receiving.
* Short message type IDs, both sending and receiving.
* Ignore packets (receiving only, but tested in a unit test).
* Session IDs are visible in `getpeerinfo` output (for manual comparison).

Things that are not included, left for future PRs, are:
* Actually using the v2 transport for connections.
* Support for the `NODE_P2P_V2` service flag.
* Retrying downgrade to V1 when attempted outbound V2 connections immediately fail.
* P2P functional and unit tests

sipa committed:
net: add V2Transport class with subset of BIP324 functionality

This introduces a V2Transport with a basic subset of BIP324 functionality:
* no ability to send garbage (but receiving is supported)
* no ability to send decoy packets (but receiving them is supported)
* no support for short message id encoding (neither encoding or decoding)
* no waiting until 12 non-V1 bytes have been received
* (and thus) no detection of V1 connections on the responder side
  (on the sender side, detecting V1 is not supported either, but that needs
  to be dealt with at a higher layer, by reconnecting)

sipa committed:
net: make V2Transport preallocate receive buffer space

sipa committed:
net: make V2Transport send uniformly random number garbage bytes

sipa committed:
net: add short message encoding/decoding support to V2Transport

Sjors commented:
Concept ACK

Mostly happy with b7a7ed70d06ab3f994ff58e3a0c99105ee88ab6e. Minor inline suggestions and/or making sure I understand what the code is doing.

I also checked that all intermediate commits compile and ran the new fuzzer for a few hours (but didn't study it very carefully).

* 1937a5fcf795149c44b7f4f016c05000ac3adaf9: `-maxsendbuffer` help could be changed to say "Maximum per-connection memory use for the send buffer"
* 68e48a0185751d24eecb194b8efd7028c8b590f3: can you elaborate a bit in the commit description why `nSendOffset` can be dropped?

sipa committed:
net: add have_next_message argument to Transport::GetBytesToSend()

Before this commit, there are only two possibly outcomes for the "more" prediction
in Transport::GetBytesToSend():
* true: the transport itself has more to send, so the answer is certainly yes.
* false: the transport has nothing further to send, but if vSendMsg has more message(s)
         left, that still will result in more wire bytes after the next
         SetMessageToSend().

For the BIP324 v2 transport, there will arguably be a third state:
* definitely not: the transport has nothing further to send, but even if vSendMsg has
                  more messages left, they can't be sent (right now). This happens
                  before the handshake is complete.

To implement this, we move the entire decision logic to the Transport, by adding a
boolean to GetBytesToSend(), called have_next_message, which informs the transport
whether more messages are available. The return values are still true and false, but
they mean "definitely yes" and "definitely no", rather than "yes" and "maybe".

Sjors commented on test/functional/test_framework/test_framework.py:
```diff
@@ -595,12 +595,12 @@ def connect_nodes(self, a, b):
         self.wait_until(lambda: sum(peer['version'] != 0 for peer in to_connection.getpeerinfo()) == to_num_peers)
-        self.wait_until(lambda: sum(peer['bytesrecv_per_msg'].pop('verack', 0) == 24 for peer in from_connection.getpeerinfo()) == from_num_peers)
-        self.wait_until(lambda: sum(peer['bytesrecv_per_msg'].pop('verack', 0) == 24 for peer in to_connection.getpeerinfo()) == to_num_peers)
+        self.wait_until(lambda: sum(peer['bytesrecv_per_msg'].pop('verack', 0) >= 21 for peer in from_connection.getpeerinfo()) == from_num_peers)
```
163d93304dea08bf8f2ea79d79cabceec1a13d96: I assume this is because of the short message ID's? Can you mention that in the commit message, so it's easy to find with `git blame`?

Sjors commented on src/net.cpp:
```diff
@@ -1434,6 +1435,20 @@ size_t V2Transport::GetSendMemoryUsage() const noexcept
+{
+    if (m_use_v1) return m_v1_fallback.GetSessionID();
+    LOCK(m_recv_mutex);
+    if (m_recv_state == RecvState::V1) return m_v1_fallback.GetSessionID();
```
9c3159378658c4b1535ab6c546512f7d1b2b3979: what's the difference between these two scenarios? `m_use_v1` vs `m_recv_state == RecvState::V1`?

Why not return `{}`? Unsafe / bad / not great?

Sjors commented:
I tend to get spontaneous v2 connections eventually, both inbound and outbound. But are there any known reachable mainnet nodes folks can test against?

sipa committed:
test: add unit tests for V2Transport

mzumsande commented on src/net.cpp:
```diff
@@ -912,6 +912,351 @@ size_t V1Transport::GetSendMemoryUsage() const noexcept
+    if (msg_type) {
+        reject_message = false;
+        msg.m_type = std::move(*msg_type);
+        msg.m_time = time;
```
In `V1Transport::GetReceivedMessage`, we call `RandAddEvent()` to harvest entropy from the time and checksum of the received messages. Should something similar be done for V2?

vincenzopalazzo commented on src/net.cpp:
```diff
@@ -912,6 +912,353 @@ size_t V1Transport::GetSendMemoryUsage() const noexcept
+            }
+            break;
+        default:
+            // Any other state is invalid (this function should not have been called).
+            assert(false);
```
maybe this is for a followup PR, but I think here it is better to print the status of the state, this makes trivial to debug the problem if happens.

vincenzopalazzo replied:
I see now your point, ok make sense

vincenzopalazzo reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/28196/commits/d9344fd5c77e69a544c2d80aa6cbd20827f2e169

Does CI failure look like an unlucky run? or unrelated


```
 node2 2023-08-29T03:35:33.769000Z [httpworker.2] [wallet/wallet.h:832] [WalletLogPrintf] [w1] AddToWallet e930f400504216afbf78e3760d4f6730547258b8a453514d7b2b5bccd2432f9a 
 node2 2023-08-29T03:35:33.769921Z [http] [httpserver.cpp:241] [http_request_cb] [http] Received a POST request for /wallet/w1 from 127.0.0.1:49136 
 node2 2023-08-29T03:35:33.769971Z [httpworker.3] [rpc/request.cpp:179] [parse] [rpc] ThreadRPCServer method=getwalletinfo user=__cookie__ 
 node2 2023-08-29T03:35:33.770827Z [http] [httpserver.cpp:241] [http_request_cb] [http] Received a POST request for /wallet/w1 from 127.0.0.1:49136 
 node2 2023-08-29T03:35:33.770859Z [httpworker.0] [rpc/request.cpp:179] [parse] [rpc] ThreadRPCServer method=listtransactions user=__cookie__ 
 test  2023-08-29T03:35:33.772000Z TestFramework (ERROR): Assertion failed 
                                   Traceback (most recent call last):
                                     File "/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 131, in main
                                       self.run_test()
                                     File "/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/wallet_backwards_compatibility.py", line 201, in run_test
                                       assert txs[3]["abandoned"]
                                   AssertionError
 test  2023-08-29T03:35:33.773000Z TestFramework (DEBUG): Closing down network thread 
```

theStack commented on src/net.cpp:
```diff
@@ -1213,7 +1276,22 @@ bool V2Transport::ReceivedBytes(Span<const uint8_t>& msg_bytes) noexcept
+        }
+    }
+
+    if (contents.size() < 13) return std::nullopt; // Long encoding needs at least 13 bytes
```
nit-like: For the long encoding processing, it might be worthwhile to advance for one byte in the contents span first, to work with the remaining 12 bytes and avoid the plus ones everywhere below (especially for the array indices in the while loops)?
```suggestion
    contents = contents.subspan(1);
    if (contents.size() < 12) return std::nullopt; // Long encoding needs at least 12 bytes (after the short message id)
```

theStack commented on src/net.cpp:
```diff
@@ -1270,9 +1348,17 @@ bool V2Transport::SetMessageToSend(CSerializedNetMsg& msg) noexcept
+        contents[0] = *short_message_id;
+        std::copy(msg.data.begin(), msg.data.end(), contents.begin() + 1);
+    } else {
+        contents.resize(13 + msg.data.size());
```
nit: though logically identical with the current code, could be explicit about the fill value of the resize (in contrast to the short message case above, not all of the new elments after the resize are overwritten):
```suggestion
        contents.resize(13 + msg.data.size(), 0);
```

theStack reviewed (commented):
Concept ACK

Left some nits regarding short message encoding below, planning to do a deeper review within the next days.

sipa committed:
net: make V2Transport auto-detect incoming V1 and fall back to it

mzumsande reviewed (commented):
Concept ACK - I reviewed only the first two commits so far in-depth.

sipa committed:
net: remove unused Transport::SetReceiveVersion

vasild commented on src/net.cpp:
```diff
@@ -910,24 +912,557 @@ size_t V1Transport::GetSendMemoryUsage() const noexcept
+        // Initialize with zeroes, and then write the message type string starting at offset 0.
+        // This means contents[0] and the unused positions in contents[1..13] remain 0x00.
+        contents.resize(1 + CMessageHeader::COMMAND_SIZE + msg.data.size(), 0);
+        std::copy(msg.m_type.begin(), msg.m_type.end(), reinterpret_cast<char*>(contents.data() + 1));
+        std::copy(msg.data.begin(), msg.data.end(), contents.begin() + 1 + CMessageHeader::COMMAND_SIZE);
```
The cast seems unnecessary?
```suggestion
        std::copy(msg.m_type.begin(), msg.m_type.end(), contents.data() + 1);
        std::copy(msg.data.begin(), msg.data.end(), contents.begin() + 1 + CMessageHeader::COMMAND_SIZE);
```

vasild reviewed (commented):
Reviewed up to 1a8892201e (incl). Posting review midway. Mostly minor stuff in the comments below.

sipa committed:
crypto: Spanify EllSwiftPubKey constructor

mzumsande commented on src/net.cpp:
```diff
@@ -912,6 +912,347 @@ size_t V1Transport::GetSendMemoryUsage() const noexcept
+            break;
+        case RecvState::VERSION:
+            if (!ignore) {
+                // Version message received; transition to application phase. The contents is
```
okay then, can be resolved.

vasild commented on src/test/net_tests.cpp:
```diff
@@ -1008,4 +1009,496 @@ BOOST_AUTO_TEST_CASE(advertise_local_address)
+    void ReceiveMessage(const std::string& m_type, Span<const uint8_t> payload)
+    {
+        auto ret = ReceivePacket();
+        BOOST_CHECK(ret.size() == payload.size() + 13);
+        BOOST_CHECK(ret[0] == 0);
+        for (unsigned i = 0; i < 12; ++i) {
+            if (i < m_type.size()) {
+                BOOST_CHECK(ret[1 + i] == m_type[i]);
+            } else {
+                BOOST_CHECK(ret[1 + i] == 0);
+            }
+        }
+        BOOST_CHECK(Span{ret}.subspan(13) == payload);
+    }
+
+    /** Schedule an encrypted packet with specified message type and payload to be sent to
+     *  transport (only after ReceiveKey). */
+    void SendMessage(std::string mtype, Span<const uint8_t> payload)
+    {
+        // Construct contents consisting of 0x00 + 12-byte message type + payload.
+        std::vector<uint8_t> contents(13 + payload.size());
```
Would be nice to avoid the magic numbers 12 and 13.

vasild commented on src/test/net_tests.cpp:
```diff
@@ -1008,4 +1009,496 @@ BOOST_AUTO_TEST_CASE(advertise_local_address)
+        BOOST_CHECK(ret && ret->empty());
+        tester.ReceiveGarbage();
+        tester.ReceiveVersion();
+        auto msg_data_1 = g_insecure_rand_ctx.randbytes<uint8_t>(InsecureRandRange(100000));
```
Since this unit test uses random numbers, it may happen that if it fails/crashes, then it does not do that if run again. How would one go to reproduce a test failure?

vasild replied:
(offtopic)

in `multiprocess, i686, DEBUG` the seed it printed: https://cirrus-ci.com/task/5770424366137344?logs=ci#L3855
```
/bin/bash: line 1: 25146 Aborted                 (core dumped) test/test_bitcoin --catch_system_errors=no -l test_suite -t "$( cat test/net_tests.cpp | grep -E "(BOOST_FIXTURE_TEST_SUITE\\(|BOOST_AUTO_TEST_SUITE\\()" | cut -d '(' -f 2 | cut -d ',' -f 1 | cut -d ')' -f 1 )" -- DEBUG_LOG_OUT > "$TEST_LOGFILE" 2>&1
...
test/net_tests.cpp(1311): Entering test case "v2transport_test"
2023-09-06T04:19:04.842541Z [test] [test/util/random.cpp:31] [Seed] Seed: Setting random seed for current tests to RANDOM_CTX_SEED=c36f0537baafdf3bc691d89186094681ace7c42292794ea4d00fa784c3b4957b
...
2023-09-06T04:19:05.182858Z [test] [net.cpp:1237] [ProcessReceivedPacketBytes] [net]net.cpp:1129 ProcessReceivedKeyBytes: Assertion `m_recv_buffer.size() <= OFFSET + MATCH.size()' failed.
```

In `previous releases, qt5 dev package and depends packages, DEBUG` it is not: https://cirrus-ci.com/task/4785261947650048?logs=ci#L3042
```
bash -c ' DIR_UNIT_TEST_DATA=/ci_container_base/ci/scratch/qa-assets/unit_test_data/ LD_LIBRARY_PATH=/ci_container_base/depends/x86_64-pc-linux-gnu/lib /ci_container_base/ci/scratch/out/bin/test_bitcoin --catch_system_errors=no -l test_suite'
...
test/net_tests.cpp(1311): Entering test case "v2transport_test"
net.cpp:1129 ProcessReceivedKeyBytes: Assertion `m_recv_buffer.size() <= OFFSET + MATCH.size()' failed.
/ci_container_base/ci/test/06_script_b.sh: line 170: 22342 Aborted                 bash -c "${TEST_RUNNER_ENV} DIR_UNIT_TEST_DATA=${DIR_UNIT_TEST_DATA} LD_LIBRARY_PATH=${DEPENDS_DIR}/${HOST}/lib ${BASE_OUTDIR}/bin/test_bitcoin --catch_system_errors=no -l test_suite"
```

@MarcoFalke do you have an idea if this is intentional? Any downsides to also use `DEBUG_LOG_OUT` in `previous releases, qt5 dev package and depends packages, DEBUG`?

vasild replied:
Opened https://github.com/bitcoin/bitcoin/issues/28466 so that this does not get forgotten.

vasild reviewed (approved):
ACK 8bd5aa1a1084c5f4c34cba85506661338e8e91ea (I need to re-review the unit test from the last commit more carefully)

Non-blocker comments below.

Sjors commented on src/net.h:
```diff
@@ -417,6 +419,187 @@ class V1Transport final : public Transport
+         * This is the initial state. The public key is sent out. When the receiver
+         * receives the other side's public key and transitions to GARB_GARBTERM, the sender state
+         * becomes KEY_GARBTERM_GARBAUTH_VERSION. The key is left in the send buffer when this
+         * happens, because it may not have been fully sent out yet. */
```
fb3ec8fd9e96f6c2d16db02fd2d683ea097a90ff: why not remain in the `KEY` state until we've completely sent out the public key? And only then transition to `GARBTERM_GARBAUTH_VERSION`?

Sjors replied:
Perhaps a more natural border would be `KEY_GARB` -> `GARBTERM_GARBAUTH_VERSION` then? But that sounds like what you implemented in the branch. Maybe just leave it as is, I might revisit this after I've looked at more of the code.

Sjors replied:
I suppose the big conceptual difference between our send and receive state machines is that the latter has a clearly defined moment when the key exchange is complete, i.e. once it received the public key bytes. From the `send` state machine point of view this happens in the background at some unpredictable point.

Sjors reviewed (commented):
Studied the state machine a bit, which makes sense to me, except the `KEY -> KEY_GARBTERM_GARBAUTH_VERSION` transition.

Running the code (using the main PR). I also looked at the fallback mechanism.

General protocol question: why are `garbage` and `decoy packets` distinct concepts in BIP324? That is, why can't garbage just be a single instance of a decoy package?

sipa reviewed (commented):
Addressed a number of comments by @vasild and @Sjors.

I've also removed the Erlay messages from the short message id list, after observing that having them there makes no difference: if peers send us these, the messages will be ignored, whether they're in the list or not. After discussing with @mzumsande I think it may also be better to remove them from the BIP for now, but it's nice to realize that they can be dropped from the code regardless of what happens to the BIP.

theStack commented on src/net.cpp:
```diff
@@ -912,6 +912,420 @@ size_t V1Transport::GetSendMemoryUsage() const noexcept
+        // Other side's key has been fully received.
+
+        // Initialize the ciphers.
+        EllSwiftPubKey ellswift(MakeByteSpan(m_recv_buffer).first(EllSwiftPubKey::size()));
```
nit: as `m_recv_buffer` has exactly the size of an ellswift pubkey at this point (ensured by the if a few lines above), the `.first` call isn't needed:
```suggestion
        EllSwiftPubKey ellswift(MakeByteSpan(m_recv_buffer));
```

theStack commented on src/net.cpp:
```diff
@@ -912,6 +912,420 @@ size_t V1Transport::GetSendMemoryUsage() const noexcept
+            return false;
+        }
+        // Feed the last 4 bytes of the Poly1305 authentication tag (and its timing) into our RNG.
+        RandAddEvent(ReadLE32(m_recv_buffer.data() + m_recv_buffer.size() - 4));
```
nit:
```suggestion
        RandAddEvent(ReadLE32(m_recv_buffer.end() - 4));
```

theStack commented on src/net.cpp:
```diff
@@ -912,6 +912,420 @@ size_t V1Transport::GetSendMemoryUsage() const noexcept
+    auto msg_type = GetMessageType(contents);
+    CDataStream ret(m_recv_type, m_recv_version);
+    CNetMessage msg{std::move(ret)};
+    msg.m_raw_message_size = m_recv_decode_buffer.size() + BIP324Cipher::EXPANSION;
```
Should the size of the 3-byte length descriptor also be added here? (at least that would come close to what happens for v1, where HEADER_SIZE is accounted for).

theStack commented on src/net.h:
```diff
@@ -417,6 +419,194 @@ class V1Transport final : public Transport
+class V2Transport final : public Transport
+{
+private:
+    /** Contents of the version packet to send. BIP324 that senders should leave this empty, and
```
missing verb after "BIP324":
```suggestion
    /** Contents of the version packet to send. BIP324 stipulates that senders should leave this empty, and
```

theStack commented on src/net.cpp:
```diff
@@ -912,6 +912,420 @@ size_t V1Transport::GetSendMemoryUsage() const noexcept
+    AssertLockHeld(m_send_mutex);
+    // No-op if no change is desired.
+    if (send_state == m_send_state) return;
+    // Enforce allowed state transitions.
+    switch (send_state) {
```
nit: in `Set{Send,Receive}State`, at least for my brain the state transition checks would be slightly easier to grasp if the current state is switch/case-d and the allowed next state(s) are `Assume`d (i.e. asserting "where can I go to?" rather than "where could I have come from?").

theStack reviewed (commented):
Reviewed up to 4810d69306ee214063386f9bfd40f89525482e8b so far, left some non-blocking nits on the way.

Sjors commented on src/net.cpp:
```diff
@@ -973,11 +987,44 @@ size_t V2Transport::GetMaxBytesToProcess() noexcept
+    // be the beginning of either a v1 packet (network magic + "version\x00"), or of a v2 public key.
+    assert(m_recv_buffer.size() <= m_v1_prefix.size());
+    if (!std::equal(m_recv_buffer.begin(), m_recv_buffer.end(), m_v1_prefix.begin())) {
+        // Mismatch with v1 prefix, so we can assume a v2 connection.
```
No strong preference here, but I imagine it's useful early on for debugging to tell the difference between a truly failed v2 connection and some other network failing to make a v1 connection. Though I have no idea how often the latter actually happens these days.

Sjors commented on src/net.h:
```diff
@@ -308,19 +308,25 @@ class Transport {
+     *            to be sent before (have_next_message=false) or after (have_next_message=true) a
+     *            potential SetMessageToSend immediately afterwards. It is set by the caller when
+     *            they know they have another message ready to send. The have_next_message
+     *            argument only affects this "more" return value and nothing else.
```
d67be9ad3836bd19cc8ea71a0b99c7d60682dd32: alternative explanation, which I think is more clear (but perhaps wrong):

> have_next_message: the caller knows it has another message ready to send. The have_next_message argument (only) affects the "more" return value, which is used to set the `MSG_MORE` TCP flag and to trigger optimistic send. With v2 transport more can be false even if vSendMsg has more messages left. This happens before the handshake is complete.

I found this part confusing: "indicates more bytes to be sent before (have_next_message=false) or after (have_next_message=true) a potential SetMessageToSend immediately afterwards"

Sjors commented on src/net.cpp:
```diff
@@ -3006,7 +3013,7 @@ void CConnman::PushMessage(CNode* pnode, CSerializedNetMsg&& msg)
     {
         LOCK(pnode->cs_vSend);
-        const auto& [to_send, _more, _msg_type] = pnode->m_transport->GetBytesToSend();
+        const auto& [to_send, more, _msg_type] = pnode->m_transport->GetBytesToSend(true);
```
d67be9ad3836bd19cc8ea71a0b99c7d60682dd32 maybe add a comment:

```
// Check if the transport still has unsent bytes,
// and indicate to it that we're about to give it a message to send.
```

`/*have_next_message=*/true`

Sjors commented on src/net.cpp:
```diff
@@ -3015,10 +3022,11 @@ void CConnman::PushMessage(CNode* pnode, CSerializedNetMsg&& msg)
         // because the poll/select loop may pause for SELECT_TIMEOUT_MILLISECONDS before actually
         // doing a send, try sending from the calling thread if the queue was empty before.
-        if (queue_was_empty) {
+        if (queue_was_empty && more) {
```
d67be9ad3836bd19cc8ea71a0b99c7d60682dd32: an alternative approach could be to have the v2 transport tell us that it's `blocked` (in the pre-handshake scenario): `&& !blocked`.

Sjors commented on src/net.cpp:
```diff
@@ -1322,9 +1328,10 @@ Sock::EventsPerSock CConnman::GenerateWaitSockets(Span<CNode* const> nodes)
+            // once a potential message from vSendMsg is handed to the transport. GetBytesToSend
+            // determines both of these in a single call.
+            const auto& [to_send, more, _msg_type] = pnode->m_transport->GetBytesToSend(!pnode->vSendMsg.empty());
+            select_send = !to_send.empty() || more;
```
d67be9ad3836bd19cc8ea71a0b99c7d60682dd32: if we were to switch to an approach of asking if the transport is blocked, this would change to:

```cpp
select_send = !to_send.empty() || (!blocked && !pnode->vSendMsg.empty());
```

Not sure if that's really more clear.

Sjors reviewed (commented):
Some thoughts on d67be9ad3836bd19cc8ea71a0b99c7d60682dd32. It might make sense to directly track if we're blocked from sending pending the handshake, instead of the indirect approach of setting `more`. But I'm not convinced that would make things easier.

drcalebpayne commented:
https://foundation.blacksintechnology.net/my-account/


On Mon, Sep 4, 2023, 3:31 PM Pieter Wuille ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/net.cpp
> <https://github.com/bitcoin/bitcoin/pull/28196#discussion_r1315218195>:
>
> > @@ -3006,7 +3013,7 @@ void CConnman::PushMessage(CNode* pnode, CSerializedNetMsg&& msg)
>      size_t nBytesSent = 0;
>      {
>          LOCK(pnode->cs_vSend);
> -        const auto& [to_send, _more, _msg_type] = pnode->m_transport->GetBytesToSend();
> +        const auto& [to_send, more, _msg_type] = pnode->m_transport->GetBytesToSend(true);
>
> Done.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/28196#discussion_r1315218195>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AVQO3WAXGHMDBSU3224QHSTXYZJCVANCNFSM6AAAAAA3AD527Y>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

naumenkogs commented on src/net.h:
```diff
@@ -417,6 +419,179 @@ class V1Transport final : public Transport
+
+        /** Garbage and garbage terminator.
+         *
+         * Whenever a byte is received, the last 16 bytes are compared with the expected garbage
```
5cb433948b050913d22914905c7db510e9ef48d2
Should be extra careful with not keeping 4kb in memory waiting here i guess... After staring at the code for a few minutes I struggle to find this logic :) So perhaps worth adding a comment for it?

This question possibly applies to the whole transport thing. Or maybe I'm just confused with how it goes through spans in the first place...

vasild replied:
Later communication is subject to ping timeout, e.g. if the peer starts sending us a new block, sends 1MB and stops any sends to us, the peer will eventually be disconnected due to ping timeout. I guess this handshake should be subject to the same. Is it? I guess that is in some later PR which will engage the V2 transport.

naumenkogs replied:
>Regarding the 4 kB, it's wiped in ProcessReceivedGarbageBytes now: when the full garbage + terminator has been processed, m_recv_buffer.clear(); is called.

What will happen if we're at 3.99 Kb of garbage and no terminator? Am i right that the caller would be responsible for terminating this after a timeout?

naumenkogs commented:
Reviewed everything 4f1900596940c7a2c7f641f2ffaa4ef51aded97a, see comments.

As for the unit tests, it would be nice to know what other coverage could be useful here. In the form of a todo or something. Or do you consider it close to complete?

vasild commented on src/net.cpp:
```diff
@@ -910,24 +912,625 @@ size_t V1Transport::GetSendMemoryUsage() const noexcept
+    AssertLockNotHeld(m_send_mutex);
+    LOCK(m_send_mutex);
+    if (m_send_state == SendState::V1) return m_v1_fallback.SetMessageToSend(msg);
+    if (m_send_state != SendState::READY) return false;
+    if (!m_send_buffer.empty()) return false;
```
nit: both conditions can be combined, they are not terribly long:
```suggestion
    if (m_send_state != SendState::READY || !m_send_buffer.empty()) return false;
```

vasild commented on src/test/net_tests.cpp:
```diff
@@ -1008,4 +1009,498 @@ BOOST_AUTO_TEST_CASE(advertise_local_address)
+    for (int i = 0; i < 10; ++i) {
+        V2TransportTester tester(true);
+        auto ret = tester.Interact();
+        BOOST_CHECK(ret && ret->empty());
```
I changed something in the test (to make it fail deliberately) and it caused a flood of error reports from boost. I am not sure how useful is this. Does it make sense to continue the test if one of these checks fails? If not, then consider replacing `BOOST_CHECK()` with `BOOST_REQUIRE()` which would stop the test on the first failure.

vasild reviewed (approved):
ACK 4f1900596940c7a2c7f641f2ffaa4ef51aded97a

Would be happy to re-review if further improvements arise from the discussions :)

Sjors reviewed (commented):
(midway review)

sipa committed:
net: detect wrong-network V1 talking to V2Transport

vasild commented on src/net.cpp:
```diff
@@ -910,24 +912,685 @@ size_t V1Transport::GetSendMemoryUsage() const noexcept
+    AssertLockNotHeld(m_send_mutex);
+    Assume(m_recv_state == RecvState::KEY);
+    Assume(m_recv_buffer.size() <= EllSwiftPubKey::size());
+    if (m_recv_buffer.size() == EllSwiftPubKey::size()) {
+        // Other side's key has been fully received, and can now be Diffie-Hellman combined with
+        // our key to initialize the encryption ciphers.
+
+        // As a special exception, if bytes 4-16 of the key on a responder connection match the
+        // corresponding bytes of a V1 version message, but bytes 0-4 don't match the network magic
+        // (if they did, we'd have switched to V1 state already), assume this is a peer from
+        // another network, and disconnect them. They will almost certainly disconnect us too when
+        // they receive our uniformly random key and garbage, but detecting this case specially
+        // means we can log it.
+        static constexpr std::array<uint8_t, 12> MATCH = {'v', 'e', 'r', 's', 'i', 'o', 'n', 0, 0, 0, 0, 0};
+        static constexpr size_t OFFSET = sizeof(CMessageHeader::MessageStartChars);
+        if (!m_initiating) {
+            Assume(m_recv_buffer.size() <= OFFSET + MATCH.size());
```
This looks wrong, the condition on the `if` on line 1116 means that `m_recv_buffer` is exactly 64 bytes. Here we assert that it is less than 4 + 12 = 16. I think this is causing the CI failure.

instagibbs reviewed (commented):
light review, comparing implementation with the BIP, did not review the tests

naumenkogs commented:
ACK 15ea0ce587a334b647119fe65de2b305720c3eba

vasild reviewed (approved):
ACK 15ea0ce587a334b647119fe65de2b305720c3eba

[Coverage report](https://people.freebsd.org/~vd/pr28196_15ea0ce587_coverage/modified_and_not_covered.html) for modified lines by this PR and not covered by tests:
```
./src/test/test_bitcoin
./src/qt/test/test_bitcoin-qt
./test/functional/test_runner.py
FUZZ=p2p_transport_bidirectional_v2   ./src/test/fuzz/fuzz -max_total_time=1800 /tmp/empty
FUZZ=p2p_transport_bidirectional_v1v2 ./src/test/fuzz/fuzz -max_total_time=1800 /tmp/empty
```

theStack commented on src/net.cpp:
```diff
@@ -912,6 +912,427 @@ size_t V1Transport::GetSendMemoryUsage() const noexcept
+    m_cipher{key, ent32}, m_initiating{initiating}, m_nodeid{nodeid},
+    m_recv_type{type_in}, m_recv_version{version_in},
+    m_recv_state{RecvState::KEY},
+    m_send_state{SendState::AWAITING_KEY}
```
follow-up consideration: the two V2Transport constructors have a lot of code in common, is it somehow possible to deduplicate their implementation (e.g. with another private constructor)?

vasild replied:
I played with this some time ago. I used a private constructor that is called by the other two public constructors (see [delegating constructor](https://stroustrup.com/C++11FAQ.html#delegating-ctor)). One problem was that `BIP324Cipher` is not movable, so I involved `std::forward` to pass the arguments from the public ctors to the private one which would construct the `BIP324Cipher` object. It worked. Also the private ctor had to have different arguments than the public ones, so I added a dummy `nullptr_t` first arg (:disappointed:). Then in a later commit of this PR the two public ctors diverged and were not identical anymore, so I decided to ditch the idea.

theStack reviewed (approved):
ACK 15ea0ce587a334b647119fe65de2b305720c3eba

Left some nitty follow-up material below.

theStack reviewed (approved):
re-ACK db9888feec48c6220a2fcf92865503bbbdab02a4

Sjors commented on src/net.cpp:
```diff
@@ -912,6 +912,418 @@ size_t V1Transport::GetSendMemoryUsage() const noexcept
+        } else if (m_recv_buffer.size() == MAX_GARBAGE_LEN + BIP324Cipher::GARBAGE_TERMINATOR_LEN) {
+            // We've reached the maximum length for garbage + garbage terminator, and the
+            // terminator still does not match. Abort.
+            LogPrint(BCLog::NET, "V2 transport error: missing garbage terminator, peer=%d\n", m_nodeid);
```
I used this in #27826, but it requires having access to the whole `CNode&`, not just `m_nodeid`:

```
fLogIPs ? strprintf(" peeraddr=%s", peer.addr.ToStringAddrPort()) : ""
```

Can wait for followup.

Sjors commented on src/net.cpp:
```diff
@@ -1327,6 +1330,40 @@ bool V2Transport::ReceivedBytes(Span<const uint8_t>& msg_bytes) noexcept
+                // to MAX_RESERVE_AHEAD above the largest allowed message contents size, and to
+                // MAX_RESERVE_AHEAD more than they've actually sent us.
+                size_t alloc_add = std::min(max_read, msg_bytes.size() + MAX_RESERVE_AHEAD);
+                m_recv_buffer.reserve(m_recv_buffer.size() + alloc_add);
```
297c8889975a18258d6cc39b1ec1e94fed6630fb: since the code worked before this commit, what happens we don't call `reserve`? Does that result in de worst case where the operating system allocates one byte at a time?

Sjors reviewed (commented):
Concept ACK.

Overall this looks pretty solid.

Commit c3fad1f29df093e8fd03d70eb43f25ee9d531bf7 looks good; there's no ideal way to do this, but at least it's now pretty well documented.

The state machine looks good too.

Otherwise I only shallowly reviewed up to db9888feec48c6220a2fcf92865503bbbdab02a4, and without looking at test and fuzz code.

To better ensure decoy packages work, maybe you could add a (debug) rpc method to send decoy package to a given peer? Though that can wait for a followup.

mzumsande reviewed (commented):
Code Review ACK db9888feec48c6220a2fcf92865503bbbdab02a4

I reviewed the code again and did a little bit of manual mutation testing (checking that the fuzz test fails if I break things).
The nits can be ignored.

naumenkogs commented:
ACK db9888feec48c6220a2fcf92865503bbbdab02a4

fanquake commented:
Outstanding/followup comments:

https://github.com/bitcoin/bitcoin/pull/28196#discussion_r1318578731
https://github.com/bitcoin/bitcoin/pull/28196#discussion_r1315179378
https://github.com/bitcoin/bitcoin/pull/28196#discussion_r1319134141

vasild reviewed (commented):
ACK db9888feec48c6220a2fcf92865503bbbdab02a4

vincenzopalazzo reviewed (commented):
re-ACK https://github.com/bitcoin/bitcoin/commit/db9888feec48c6220a2fcf92865503bbbdab02a4
